portal-kernel/src/com/liferay/portal/kernel/portlet/PortletContainerUtil.java

```diff
@@ -157,7 +158,62 @@ public static void processAction(
 							host.equals(renderURL.getHost()) &&
 							(port == renderURL.getPort())) {
 
-							location = renderURL.toString();
+							String renderURLQuery = renderURL.getQuery();
+
+							if (renderURLQuery != null) {
+								StringBundler fileSB = new StringBundler();
+
+								if (locationURL.getPath() != null) {
+									fileSB.append(locationURL.getPath());
+								}
+
+								fileSB.append("?");
+
+								boolean firstParam = true;
+
+								String query = locationURL.getQuery();
+
+								if (query != null) {
+									fileSB.append(query);
+									firstParam = false;
+								}
+
+								String publicRenderParameterNamespace =
+									PortletQName.
+										PUBLIC_RENDER_PARAMETER_NAMESPACE;
+
+								String[] renderURLParams = renderURLQuery.split(
+									"[&]");
+
+								for (String nameValuePair : renderURLParams) {
+									String name = nameValuePair.substring(
+										0, nameValuePair.indexOf("="));
+
+									if (name.startsWith(
+											publicRenderParameterNamespace) &&
+										!query.contains(name)) {
```

Comment:
!query.contains(name), looks wrong to me.

What if we have something like:
renderURLQuery = ?p_r_p_name=value1
query = p_r_p_name2=value2

Then "name" is contained in query even though it's not a parameter.  Also you have a null check on query but in the loop it doesn't have a null check, so either there's a possible NPE or a useless null check.

Reply:
Also note the contains could be wrong the other way, could be something like p_r_p_1name.

Reply:
@Preston-Crary Good catch. I'll make a correction and get back to you.

Reply:
See https://github.com/Preston-Crary/liferay-portal/pull/614